### ISSUE TITLE
Check if Atomic CSS affects Shadow DOM

### ISSUE THREAD

**Issue by miroslavstastny** [microsoft/fluentui#24948]
⬇️ written by @layershifter 

---

TL;DR On mount it's _slower_ to use Shadow DOM.

---

To test some assumptions I created [a repo with different scenarios](https://github.com/layershifter/test-atomic-shadow-dom) to compare mount performance in real DOM and Shadow DOM. For Shadow DOM I used [constructable stylesheets](https://web.dev/constructable-stylesheets/) and insertion directly to a style tag.

The idea of a benchmark is to render 10k elements (each has 100 CSS rules applied) and compare their behavior. I got following results ⬇️

#### Real DOM / Atomic CSS (Chrome profiler)

![image](https://user-images.githubusercontent.com/14183168/192265334-9cefe209-78f0-42e3-9523-63889f776966.png)

#### Shadow DOM / Atomic CSS (Chrome profiler)

![image](https://user-images.githubusercontent.com/14183168/192265540-bbdd86b3-6a6e-4982-8f79-f61204cdb056.png)

Noticeable things:
- Real DOM has a lot of work spent in "Recalculate style" & "Layout"
- Shadow DOM has significantly more work in JS scripting, but "Recalculate style" & "Layout" are absent 😳

The last fact is strange and something that I could not believe, so next tests were performed using [Perfetto UI](https://perfetto.dev/) that allows to trace & analyze low level actions better.

#### Shadow DOM / Atomic CSS (constructable stylesheets)
`Document::recalcStyle` takes 118ms 299us

![image](https://user-images.githubusercontent.com/14183168/192266270-3b4d701a-80d5-4eb8-b782-e406bebc5238.png)

#### Shadow DOM / Atomic CSS (style tag)
`Document::recalcStyle` takes 118ms 551us

![image](https://user-images.githubusercontent.com/14183168/192266243-c631998f-04dc-4aab-8498-582a20f5286f.png)

#### Real DOM / Atomic CSS
`Document::recalcStyle` takes 80ms 27us

![image](https://user-images.githubusercontent.com/14183168/192266171-5cd44bfe-9398-421d-95e5-0e2b5bc8eefe.png)

#### Shadow DOM / Monolthic CSS (contructable stylesheets)

`Document::recalcStyle` is 22ms 462us

![image](https://user-images.githubusercontent.com/14183168/192266530-3816e737-87ec-418d-b0ad-4a40259e5d95.png)

### Conclusions

- "Recalculate style" & "Layout" with Atomic CSS are more expensive than for monolithic CSS i.e. 100 classes on an element are worse than 1 class.
  - This statement is true for both real DOM & usages inside Shadow DOM.
- For "Recalculate style" it does not matter how we insert styles (via `<style />` or constructable stylesheets).
- Shadow DOM has additional price due `StyleEngine::updateActiveStyleSheets` called for shadow roots.
